### PR TITLE
Add welcome for boosters

### PR DESCRIPTION
**Pull request by tenhobi**

### CODE REVIEW

File: src/HonzaBotner.Discord.Services/EventHandlers/ServerBoosterHandler.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        {
+            _logger = logger;
+            _commonOptions = commonOptions.Value;
+        }
+
+        public async Task<EventHandlerResult> Handle(GuildUpdateEventArgs eventArgs)
+        {
+            if ((eventArgs.GuildBefore.PremiumSubscriptionCount ?? 0) <=
+                (eventArgs.GuildAfter.PremiumSubscriptionCount ?? 0))
+                return EventHandlerResult.Continue;
```
Comment (ostorc) [inline src/HonzaBotner.Discord.Services/EventHandlers/ServerBoosterHandler.cs]: Maybe add check to see if the subscriber count is lower after

File: src/HonzaBotner.Discord.Services/EventHandlers/ServerBoosterHandler.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                }
+            }
+
+            HashSet<ulong> newSubscribers = subscribersAfter.Except(subscribersBefore).ToHashSet();
+
+            if (!newSubscribers.Any())
+            {
+                _logger.LogError("No new subscriber even while count after is greater than count before");
+                return EventHandlerResult.Continue;
+            }
```
Comment (ostorc) [inline src/HonzaBotner.Discord.Services/EventHandlers/ServerBoosterHandler.cs]: Do we need this, cant we just use the check by counting them?

File: src/HonzaBotner.Discord.Services/EventHandlers/ServerBoosterHandler.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+            foreach (ulong newSubscriberId in newSubscribers)
+            {
+                try
+                {
+                    DiscordMember newSubscriber = await eventArgs.GuildAfter.GetMemberAsync(newSubscriberId);
+                    DiscordChannel channel = eventArgs.GuildAfter.GetChannel(_commonOptions.GentlemenChannelId);
+                    await channel.SendFileAsync(
+                        _commonOptions.GentlemenFilePath,
+                        $"Šálí mě snad zrak? Další gentleman! Přidal se k nám <@{newSubscriber.Id}>!"
```
Comment (ostorc) [inline src/HonzaBotner.Discord.Services/EventHandlers/ServerBoosterHandler.cs]: Cringe AF

**Review by ostorc — DISMISSED** [review]

File: src/HonzaBotner.Discord.Services/EventHandlers/ServerBoosterHandler.cs (hunk 3 above)
Comment (tenhobi) [reply src/HonzaBotner.Discord.Services/EventHandlers/ServerBoosterHandler.cs]: 😂 It's super dope. 😂 

Anyway, any ideas to improve it?

File: src/HonzaBotner.Discord.Services/EventHandlers/ServerBoosterHandler.cs (hunk 1 above)
Comment (tenhobi) [reply src/HonzaBotner.Discord.Services/EventHandlers/ServerBoosterHandler.cs]: What is the difference in contrast to this?

File: src/HonzaBotner.Discord.Services/EventHandlers/ServerBoosterHandler.cs (hunk 2 above)
Comment (tenhobi) [reply src/HonzaBotner.Discord.Services/EventHandlers/ServerBoosterHandler.cs]: The subscription count is lower in this part, but this check only checks we successfully collected all the new newly subscribed members (probably always 1).

But yes, we can commit this. It should work.

File: src/HonzaBotner.Discord.Services/EventHandlers/ServerBoosterHandler.cs (hunk 1 above)
Comment (ostorc) [reply src/HonzaBotner.Discord.Services/EventHandlers/ServerBoosterHandler.cs]: Oh my bad, I thought it was ==....

File: src/HonzaBotner/appsettings.BotDev.json [hunk 4]
```diff
@@ -21,7 +21,9 @@
         "VerificationMessageId": 811282731195432991,
         "VerificationChannelId": 811282707215679569,
         "VerificationEmojiName": "⚡",
-        "StaffVerificationEmojiName": "👑"
+        "StaffVerificationEmojiName": "👑",
+        "GentlemenChannelId": 0,
+        "GentlemenFilePath": "../../docs/gentleman.gif"
```
Comment (ostorc) [inline src/HonzaBotner/appsettings.BotDev.json]: Have you verified that the file is added to the image when building in heroku?

**Review by ostorc — DISMISSED** [review]
If you have verified the file, then you can merge it.

**Comment by tenhobi** [comment]
Discussion: It occurred to me, wouldn't it be better rather detect Server Booster role addition, rather than this? We can process it and "test it in production" better. 😉

Also, we can probably make one generic `WelcomeHandler` that would handle all sorts of these types of events for new dirty coders, server boosters, etc. (to reduce the number of single method handlers.

**Comment by tenhobi** [comment]
Hmm, not sure if we want to keep it merged like this. 🤔 If so, why not merge auth and staff auth, etc?

File: src/HonzaBotner.Discord.Services/EventHandlers/WelcomeRoleHandler.cs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+                _logger.LogWarning(e, "Couldn't get horny jail role");
+            }
+
+            try
+            {
+                DiscordChannel channel = eventArgs.Guild.GetChannel(_commonOptions.HornyJailChannelId);
+                await channel.SendFileAsync(
+                    _commonOptions.HornyJailFilePath,
+                    $"Ajaj, <@{eventArgs.Member.Id}>, další dirty coder!"
+                );
```
Comment (ostorc) [inline src/HonzaBotner.Discord.Services/EventHandlers/WelcomeRoleHandler.cs]: This part looks quite similar to the booster handler... Maybe extract one method, which would do all the work and you would just call it with right parameters?

**Review by ostorc — DISMISSED** [review]

File: src/HonzaBotner.Discord.Services/EventHandlers/Helpers/WelcomeMessage.cs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+using DSharpPlus.Entities;
+using DSharpPlus.EventArgs;
+using HonzaBotner.Discord.EventHandler;
+using Microsoft.Extensions.Logging;
+
+namespace HonzaBotner.Discord.Services.EventHandlers.Helpers
+{
+    public abstract class WelcomeMessage
```
Comment (ostorc) [inline src/HonzaBotner.Discord.Services/EventHandlers/Helpers/WelcomeMessage.cs]: I think name of this class should be WelcomeMessageHandler.

**Review by ostorc — CHANGES_REQUESTED** [review]

**Pull request merged**